### PR TITLE
Vanish improvements

### LINKED ISSUE #5: When I vanish, I am still visible in the server listing
Is there any way for this to be fixed?

### PR DESCRIPTION
**Pull request by samolego**
This commit should fix issue #5 and #7 . It's a simple mixin injection to the write method of the query packet class.

As seen in the picture, there are actually 2 players online, but Alex is vanished.
showcase

Should the server ping show 1/2 players online (player is still kicked upon joining since it's actually full) or should the max player count be modified as well? (So it would show 1/1 players online on the client)

### CODE REVIEW

**Comment by samolego** [comment]
I was thinking more of the max player count number; maybe it would be best to show the actual number of max players if the server is not full. Otherwise, show the modified max count.

**Comment by ghost** [comment]
Would be nice to be able to configure gunpowder so that vanished players do not to count toward the max player count, so that cheaters would not get suspicious over a changing max player count.

**Comment by samolego** [comment]
So you mean it would allow more players than the max count (if we include vanished ones)?

What happens when a player disables vanish mode? It would then show x/y players on client, x being > y ...

**Comment by ghost** [comment]
> 
> 
> So you mean it would allow more players than the max count (if we include vanished ones)?
> 
> What happens when a player disables vanish mode? It would then show x/y players on client, x being > y ...

This can happen in vanilla if an OP has the bypass player limit set to "true" in ops.json.

**Comment by samolego** [comment]
>This can happen in vanilla if an OP has the bypass player limit set to "true" in ops.json.

Ah, ok, wasn't aware of this. I'll have another look into it then.

**Comment by samolego** [comment]
This is drafted again, since #7 needs to be fixed before this is merged. Otherwise, if player leaves while vanished, the vanished count isn't decreased.

**Comment by samolego** [comment]
This is now ready to be merged.

To sum it up:
Fixes #5  and #7 by keeping the track of the vanished player count.

**Comment by samolego** [comment]
@i509VCB sorry for ping, but could you review this? Martmists said he's kinda burned out atm.
Thanks.

File: src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java [hunk 1]
```diff
@@ -24,23 +24,69 @@
...
+                    value = "INVOKE",
+                    target = "Lnet/minecraft/server/PlayerManager;broadcastChatMessage(Lnet/minecraft/text/Text;Lnet/minecraft/network/MessageType;Ljava/util/UUID;)V",
+                    shift = At.Shift.BEFORE
+            ),
+            cancellable = true
+    )
+    private void noVanishDisconnectMessage(Text reason, CallbackInfo ci) {
+        if(((PlayerVanish) player).isVanished()) {
+            PlayerManager playerManager = this.server.getPlayerManager();
+            ((VanishedPlayerManager) playerManager).setVanishedCount(((VanishedPlayerManager) playerManager).getVanishedCount() - 1);
```
Comment (TheEpicBlock) [inline src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java]: You might as well cast `playerManager` to a `VanishedPlayerManager` directly. 
Besides that the pr looks fine to me :D

**Review by TheEpicBlock — APPROVED** [review]

**Comment by i509VCB** [comment]
@samolego if you poke me tomorrow in discord I can look then. I'd say in about 16 hours?

File: src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java (hunk 1 above)
Comment (samolego) [reply src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java]: Oh, you're right. Will change that.
Thanks!

**Comment by samolego** [comment]
> @samolego if you poke me tomorrow in discord I can look then. I'd say in about 16 hours?

Ok, thank you. 11 hrs from now on :-)

File: src/main/java/io/github/gunpowder/mixin/utilities/PlayerManagerMixin_Utilities.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            method = "checkCanJoin(Ljava/net/SocketAddress;Lcom/mojang/authlib/GameProfile;)Lnet/minecraft/text/Text;",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraft/server/PlayerManager;canBypassPlayerLimit(Lcom/mojang/authlib/GameProfile;)Z",
+                    shift = At.Shift.BEFORE
+            ),
+            cancellable = true
+    )
+    private void checkCanJoinServer(SocketAddress address, GameProfile profile, CallbackInfoReturnable<Text> cir) {
+        PlayerManager manager = GunpowderMod.getInstance().getServer().getPlayerManager();
```
Comment (i509VCB) [inline src/main/java/io/github/gunpowder/mixin/utilities/PlayerManagerMixin_Utilities.java]: Wouldn't this when mixed into PlayerManager just mean this = PlayerManager?

File: src/main/java/io/github/gunpowder/mixin/utilities/PlayerManagerMixin_Utilities.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public int getVanishedCount() {
+        return this.vanishedCount;
+    }
+
+    @Override
+    public void setVanishedCount(int vanishedCount) {
+        this.vanishedCount = vanishedCount;
+    }
+
+
```
Comment (i509VCB) [inline src/main/java/io/github/gunpowder/mixin/utilities/PlayerManagerMixin_Utilities.java]: Quite a bit of whitespace, probably drop to 1 or 0 empty lines

File: src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java [hunk 4]
```diff
@@ -24,23 +24,69 @@
...
         //    ci.cancel();
         //}
     }
+
+    @Inject(
+            method = "onDisconnected(Lnet/minecraft/text/Text;)V",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraft/server/PlayerManager;broadcastChatMessage(Lnet/minecraft/text/Text;Lnet/minecraft/network/MessageType;Ljava/util/UUID;)V",
+                    shift = At.Shift.BEFORE
```
Comment (i509VCB) [inline src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java:75]: Doesn't an Inject already target the opcode before calling `broadcastChatMessage`? I don't think the extra shift is needed

File: src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java [hunk 5]
```diff
@@ -24,23 +24,69 @@
...
+            ),
+            cancellable = true
+    )
+    private void noVanishDisconnectMessage(Text reason, CallbackInfo ci) {
+        if(((PlayerVanish) player).isVanished()) {
+            VanishedPlayerManager vanishedPlayerManager = (VanishedPlayerManager) this.server.getPlayerManager();
+            vanishedPlayerManager.setVanishedCount(vanishedPlayerManager.getVanishedCount() - 1);
+            this.player.onDisconnect();
+            this.server.getPlayerManager().remove(this.player);
+            if (this.server.isHost(this.player.getGameProfile())) {
```
Comment (i509VCB) [inline src/main/java/io/github/gunpowder/mixin/utilities/ServerPlayNetworkHandlerMixin_Utilities.java:85]: obviously we can't do much here about duplicate code from vanilla, unless we want to start redirecting crap which could definitely conflict.

Honestly not an issue right now but I do think fabric api needs to have more precise disconnect/join events to handle the display of messages, but for a later day,

**Comment by samolego** [comment]
This is ready to be merged now I guess 😃 ?

**Review by i509VCB — APPROVED** [review]
Of course I assume you gave a real client and server a run to make sure this works.

**Comment by samolego** [comment]
> Of course I assume you gave a real client and server a run to make sure this works.

Yep, it was tested in singleplayer(+ LAN) and server

File: src/main/kotlin/io/github/gunpowder/mixin/cast/VanishedPlayerManager.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.github.gunpowder.mixin.cast
+
+interface VanishedPlayerManager {
+    var vanishedCount: Int
```
Comment (hYdos) [inline src/main/kotlin/io/github/gunpowder/mixin/cast/VanishedPlayerManager.kt:28]: im sorry, i dont speak bad JVM language

**Comment by Martmists-GH** [comment]
@i509VCB if you could give this one last review

**Comment by samolego** [comment]
Sorry I was trying to rename branch to open another PR to fix #17 and messed things up

File: src/main/kotlin/io/github/gunpowder/mixin/cast/VanishedPlayerManager.kt (hunk 6 above)
Comment (mounderfod) [reply src/main/kotlin/io/github/gunpowder/mixin/cast/VanishedPlayerManager.kt:28]: > im sorry, i dont speak bad JVM language

this is brilliant

**Pull request merged**